packages/data/src/components/with-dispatch/index.js

```diff
@@ -1,138 +1,97 @@
  * ```
  *
  * @example
- * In the majority of cases, it will be sufficient to use only two first params passed to `mapDispatchToProps` as illustrated in the previous example. However, there might be some very advanced use cases where using the `registry` object might be used as a tool to optimize the performance of your component. Using `select` function from the registry might be useful when you need to fetch some dynamic data from the store at the time when the event is fired, but at the same time, you never use it to render your component. In such scenario, you can avoid using the `withSelect` higher order component to compute such prop, which might lead to unnecessary re-renders of your component caused by its frequent value change. Keep in mind, that `mapDispatchToProps` must return an object with functions only.
+ * In the majority of cases, it will be sufficient to use only two first params
+ * passed to `mapDispatchToProps` as illustrated in the previous example.
+ * However, there might be some very advanced use cases where using the
+ * `registry` object might be used as a tool to optimize the performance of
+ * your component. Using `select` function from the registry might be useful
+ * when you need to fetch some dynamic data from the store at the time when the
+ * event is fired, but at the same time, you never use it to render your
+ * component. In such scenario, you can avoid using the `withSelect` higher
+ * order component to compute such prop, which might lead to unnecessary
+ * re-renders of your component caused by its frequent value change.
+ * Keep in mind, that `mapDispatchToProps` must return an object with functions
+ * only.
  *
  * ```jsx
  * function Button( { onClick, children } ) {
- * 	return <button type="button" onClick={ onClick }>{ children }</button>;
+ *     return <button type="button" onClick={ onClick }>{ children }</button>;
  * }
  *
  * const { withDispatch } = wp.data;
  *
  * const SaleButton = withDispatch( ( dispatch, ownProps, { select } ) => {
- * 	// Stock number changes frequently.
- * 	const { getStockNumber } = select( 'my-shop' );
- * 	const { startSale } = dispatch( 'my-shop' );
- *
- * 	return {
- * 		onClick() {
- * 			const dicountPercent = getStockNumber() > 50 ? 10 : 20;
- * 			startSale( discountPercent );
- * 		},
- * 	};
+ *    // Stock number changes frequently.
+ *    const { getStockNumber } = select( 'my-shop' );
+ *    const { startSale } = dispatch( 'my-shop' );
+ *    return {
+ *        onClick() {
+ *            const discountPercent = getStockNumber() > 50 ? 10 : 20;
+ *            startSale( discountPercent );
+ *        },
+ *    };
  * } )( Button );
  *
  * // Rendered in the application:
  * //
  * //  <SaleButton>Start Sale!</SaleButton>
  * ```
- * _Note:_ It is important that the `mapDispatchToProps` function always returns an object with the same keys. For example, it should not contain conditions under which a different value would be returned.
```

Comment:
Do you not foresee it being possible in the future for optimizations which leverage the fact we'd established an expectation that the prop names don't change? The previous implementation benefitted by the fact we could compute the proxy functions _once_ for the entire lifecycle of a component. I'm not sure if or how it would make sense to reflect that in the new implementation, but changing the documentation here may risk eliminating this as an option to explore down the line.

Reply:
Oh, I removed this because I thought it was documentation _I_ had added in iterating on this pull!  I agree we should keep this in place.  I'll add back in a separate pull (might not be till later)

Reply:
see https://github.com/WordPress/gutenberg/pull/15987